### PR TITLE
PUD-1281: Delete a last known address.

### PR DESCRIPTION
**Pull request by philsegal**
- add support for /recalls/{recallId}/* for MissingDocumentsRecords & LastKnownAddresses

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/LastKnownAddressController.kt [hunk 1]
```diff
@@ -37,21 +41,31 @@ class LastKnownAddressController(
...
+        ExampleProperty(
+          mediaType = "application/json",
+          value = "{\n\"status\": 404,\n\"message\":\"RecallNotFoundException(recallId=...)\"\n}"
+        )
+      )
     )
   )
-  @PostMapping("/last-known-addresses")
+  @PostMapping(
+    "/recalls/{recallId}/last-known-addresses",
```
Comment (AndyNewton) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/LastKnownAddressController.kt:53]: OK; as per thread, not sure we need this but 🤷‍♂️

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/LastKnownAddressComponentTest.kt [hunk 2]
```diff
@@ -109,9 +111,33 @@ class LastKnownAddressComponentTest : ComponentTestBase() {
...
     val expectedStatus = HttpStatus.NOT_FOUND
-    val response = authenticatedClient.lastKnownAddress(createLastKnownAddressRequest, expectedStatus, ErrorResponse::class.java)
+    val response = authenticatedClient.addLastKnownAddress(notFoundRecallId, createLastKnownAddressRequest, expectedStatus, ErrorResponse::class.java)
 
     assertThat(response, present())
     assertThat(response, equalTo(ErrorResponse(expectedStatus, "RecallNotFoundException(recallId=$notFoundRecallId)")))
   }
+
+  @Test
+  @Transactional
```
Comment (AndyNewton) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/LastKnownAddressComponentTest.kt:121]: We're not using `@Transactional` on any other component tests that I can see ... so why here?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/integration/db/LastKnownAddressRepositoryIntegrationTest.kt [hunk 3]
```diff
@@ -103,4 +105,37 @@ class LastKnownAddressRepositoryIntegrationTest : IntegrationTestBase() {
...
     OffsetDateTime.now()
   )
+
+  @Test
+  @Transactional
+  fun `deleting a last known address that exists for recall`() {
+    val lastKnownAddressId = ::LastKnownAddressId.random()
+    recallRepository.save(recall, currentUserId)
+
+    lastKnownAddressRepository.saveAndFlush(lastKnownAddress(lastKnownAddressId, 1))
```
Comment (AndyNewton) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/integration/db/LastKnownAddressRepositoryIntegrationTest.kt:115]: Just 🤔  whether `saveAndFlush` allows for the retrieve via the recall to work in these tests too?
And whether we should move to `saveAndFlush` in place of `save` for all tests in this directory - with @Transactional on them all too, I'd assume.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/integration/db/LastKnownAddressRepositoryIntegrationTest.kt [hunk 4]
```diff
@@ -103,4 +105,37 @@ class LastKnownAddressRepositoryIntegrationTest : IntegrationTestBase() {
...
+      lastKnownAddressRepository.deleteByRecallIdAndLastKnownAddressId(recall.recallId(), lastKnownAddressId)
+    }
+  }
+
+  @Test
+  @Transactional
+  fun `deleting a last known address that exists for different recall throws exception`() {
+    val lastKnownAddressId = ::LastKnownAddressId.random()
+    recallRepository.save(recall, currentUserId)
+    lastKnownAddressRepository.save(lastKnownAddress(lastKnownAddressId, 1))
```
Comment (AndyNewton) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/integration/db/LastKnownAddressRepositoryIntegrationTest.kt:135]: Happy case test uses `lastKnownAddressRepository.saveAndFlush` so preferable here too IMO.

**Review by AndyNewton — CHANGES_REQUESTED** [review]
Couple of test tweaks please.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/integration/db/LastKnownAddressRepositoryIntegrationTest.kt (hunk 3 above)
Comment (philsegal) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/integration/db/LastKnownAddressRepositoryIntegrationTest.kt:115]: nope. It didn't make a difference. The only place it seems to is when we create recalls with documents.

**Review by AndyNewton — APPROVED** [review]

**Pull request merged**